### PR TITLE
Add the kubectl apply command

### PR DESCRIPTION
**Pull request by jackgr**
This pull request adds the kubectl apply command. It does not yet create configuration annotations automatically when configuration is updated through other code paths. However, it does create them when it applies new configurations, so two or more invocations of kubectl apply will bootstrap them.

The next and final step in implementing #1702 will be automatically creating configuration annotations when configuration is updated through other code paths.

### CODE REVIEW

**Comment by jackgr** [comment]
Note that this pull request depends on #13820, which adds the method to create the three way merge patch used by kubectl apply.

**Comment by jackgr** [comment]
cc @bgrant0607, @ghodss, @mikedanese

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl

File: pkg/kubectl/cmd/apply.go [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+const (
+	apply_long = `Apply a configuration to a resource by filename or stdin.
+
+JSON and YAML formats are accepted.`
+	apply_example = `# Apply the configuration in pod.json to a pod.
+$ kubectl apply -f ./pod.json
+
+# Apply the JSON passed into stdin to a pod.
+$ cat pod.json | kubectl apply -f -`
+	apply_annotation = "k8s-kubectl-apply"
```
Comment (bgrant0607) [inline pkg/kubectl/cmd/apply.go]: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/api-conventions.md#label-selector-and-annotation-conventions

Use something like `kubectl.kubernetes.io/config`

See also the annotations used by rolling update:
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/rolling_updater.go#L38
https://github.com/kubernetes/kubernetes/blob/77a06cbf0521c6513bc6062cd0671564124c54ef/pkg/kubectl/kubectl.go#L27

**Comment by bgrant0607** [comment]
Will need an e2e test.

File: pkg/kubectl/cmd/apply.go (hunk 1 above)
Comment (jackgr) [inline pkg/kubectl/cmd/apply.go]: Done.

**Comment by jackgr** [comment]
@bgrant0607, @ghodss Added e2e test. PTAL.

**Comment by jlowdermilk** [comment]
I left review comments on the second commit. BTW, please rename second commit message; they're currently both called the same thing.

**Comment by jackgr** [comment]
@jlowdermilk Done. PTAL.

File: pkg/kubectl/cmd/apply.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	}
+
+	mapper, typer := f.Object()
+	r := resource.NewBuilder(mapper, typer, f.ClientMapperForCommand()).
+		Schema(schema).
+		ContinueOnError().
+		NamespaceParam(cmdNamespace).DefaultNamespace().
+		FilenameParam(enforceNamespace, options.Filenames...).
+		Flatten().
+		Latest().
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/apply.go]: Remove Latest() here. The reason is that the resource builder will overwrite `info.Object` when you call `Visit` with the latest version from server. You want to save/serialize the object as loaded from file *before* you get the latest version of the object in order to do a 3-way patch. See suggested changes below.

File: pkg/kubectl/cmd/apply.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	count := 0
+	err = r.Visit(func(info *resource.Info, err error) error {
+		// In this method, info.Object contains the object retrieved from the server
+		// and info.VersionedObject contains the object decoded from the input source.
+		if err != nil {
+			return err
+		}
+
+		// Serialize the current configuration of the object from the server.
+		current, err := info.Mapping.Codec.Encode(info.Object)
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/apply.go]: Per above, remove `Latest()` and change the order you serialize `current`, `original`, `modified` to
```
// Do this BEFORE calling info.Get()
modified, err := kubectl.GetModifiedConfiguration(info, true)
if err != nil {
		...
}

// Now call info.Get() to retrieve the latest version
if err := info.Get(); err != nil {
   ...
}

current, err := info.Mapping.Codec.Encode(info.Object)
if err != nil {
		...
}

// Now, retrieve original configuration from the latest server-side version of the object
original, err := kubectl.GetOriginalConfiguration(info)
if err != nil {
		...
}

// Compute a three way strategic merge patch ...
```

File: pkg/kubectl/cmd/apply.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		// Get the modified configuration of the object. Embed the result
+		// as an annotation in the modified configuration, so that it will appear
+		// in the patch sent to the server.
+		modified, err := kubectl.GetModifiedConfiguration(info, true)
+		if err != nil {
+			return cmdutil.AddSourceToErr(fmt.Sprintf("retrieving modified configuration from:\n%v\nfor:", info), info.Source, err)
+		}
+
+		// Compute a three way strategic merge patch to send to server.
+		patch, err := strategicpatch.CreateThreeWayMergePatch(original, modified, current, info.VersionedObject, false)
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/apply.go]: Another thing : did you intend to do the 3way merge patch with the "original" annotation still in the object? I would think the correct procedure would be to do the threeway merge patch without the `latest-applied-configuration`, then insert it into the patch afterwards. i.e.
1. extract `original` from `current`'s annotation
2. delete annotation from `current`
3. threewaymergepatch original, modified, current
4. add annotation = modified to the patch

Could also be you worked thru this case and merging the annotation is correct. If so, it deserves a comment explaining why.

Comment (jlowdermilk) [inline pkg/kubectl/cmd/apply.go]: Sorry, please disregard this comment, I see #14626 deals with updating the annotation.

Comment (jackgr) [inline pkg/kubectl/cmd/apply.go]: The new annotation in the patch will overwrite the previous one. That should be clear from the semantics of 3 way merge patch, per the comments there, but I'll leave another comment here to make sure the reader understands.

File: pkg/kubectl/cmd/apply.go (hunk 3 above)
Comment (jackgr) [inline pkg/kubectl/cmd/apply.go]: All of the code works correctly today. I'd like to better understand what you're asking for before changing anything. I'll set up a meeting to walk through it together.

File: pkg/kubectl/cmd/apply.go (hunk 2 above)
Comment (jackgr) [inline pkg/kubectl/cmd/apply.go]: This code works correctly as currently written. Without Latest(), it did not fetch from the server. With Latest(), it does, but does not overwrite the Versioned Object, which is what we use to diff against the annotation.

In other words, this code works correctly as currently written. I'd like to better understand what you're asking for before changing anything.

**Comment by jackgr** [comment]
Not sure how, but the edits I made yesterday got lost during the rebase. Restored them and resubmitted.

cc @jlowdermilk

**Comment by jlowdermilk** [comment]
lgtm

**Comment by jackgr** [comment]
@jlowdermilk All issues have been addressed. Can you give it an LGTM?

**Comment by piosz** [comment]
Removing LGTM because there was a change after LGTM was granted. @jlowdermilk could you please take a look?

**Pull request merged**